### ISSUE TITLE
Can't auto eager load has_many through association with a scope that uses includes

### ISSUE THREAD

**Issue by jturkel** [salsify/goldiloader#14]
Consider the following models:

```
class Blog < ActiveRecord::Base
  has_many :posts
  has_many :authors, -> { includes(:address).where('addresses.city IS NOT NULL').references(:address) }, through: :posts
end

class Post < ActiveRecord::Base
  belongs_to :blog
  belongs_to :author
end

class Author < ActiveRecord::Base
  has_many :posts
  has_one :address
end

class Address < ActiveRecord::Base
  belongs_to :author
end
```

Attempting to auto eager load (or regular eager load) `Blog#authors` results in the following error:

```
ActiveRecord::ConfigurationError: Association named 'address' was not found on Post; perhaps you misspelled it?
```

Underlying Rails bug is fixed by https://github.com/rails/rails/pull/12725. Workaround in Golidloader is to set `auto_include = false`.

**Comment by jturkel** [comment]
I've confirmed this problem still exists in Rails master. Reproducible test case can be found here: https://gist.github.com/jturkel/fc4cca3949724da4b902

**Comment by jturkel** [comment]
Looks like there's a new Rails PR up to fix this issue: https://github.com/rails/rails/pull/23100

**Comment by NullVoxPopuli** [comment]
is this still an issue?

**Comment by jturkel** [comment]
@NullVoxPopuli - Yes, it's still an issue. The Rails bug still seems to be present in Rails 5.1.2. See this [gist](https://gist.github.com/jturkel/84c50e76c1a6cc074d172ade4c97c19b).

**Comment by SzNagyMisu** [comment]
Encountered the same problem (ruby 2.4.1, Rails 5.1.4). Only I do not have any `includes` in the scope

    class Blog < ApplicationRecord
      has_many :posts
      has_many :authors, -> { where name: 'James' }, through: :posts
    end

    class Post < ApplicationRecord
      belongs_to :author
      belongs_to :blog
    end

    class Author < ApplicationRecord
      has_many :posts
    end

Of course, I started with something more complex in the scope, only when debugging, I found, that it didn't really matter what I put in there as long as it shows up in the where clause...

So, when I query `Blog.includes(:authors)`, it tries the following lines of SQL:

    SELECT `blogs`.* FROM `blogs`
    SELECT `posts`.* FROM `posts` WHERE `authors`.`name` = 'James'
      AND `posts`.`bog_id` IN (1, 2)

Where the clause `authors.name = 'James'` does not make sense for mysql, since it made no queries to the table `authors`.

Also, if I remove the scope from `has_many :authors`, the query above runs the same two line (except for the where clause, of course), and the `authors` table is not even queried until I really want something with them (like `Blog.includes(:authors).map { |blog| blog.authors.map &:name }`). Then it runs one extra query:

    SELECT `authors`.* FROM `authors` WHERE `authors`.`id` IN (1, 2)

**Lastly**, if I put the scope on the join association

    class Blog < ApplicationRecord
      has_many :posts
      has_many :posts_of_james, -> { joins(:author).where(authors: {name: 'James'}) }, class_name: 'Post'
      has_many :authors, through: :posts_of_james
    end

it runs the inner join query missing from the first approach.

    SELECT `posts`.* FROM `posts` INNER JOIN `authors`
      ON `authors`.`id` = `posts`.`author_id`
      WHERE (`authors`.`name` = 'James') AND `posts`.`id` IN (1, 2)
    SELECT `authors`.* FROM `authors` WHERE `authors`.`id` IN (1, 2)

**Comment by jturkel** [comment]
This has been fixed in Rails 5.2. See this [gist](https://gist.github.com/jturkel/8693bd3e07db2cbe92b034677dfa2491) for a reproducible test case.
